activation/poet_client_test.go

```diff
@@ -46,15 +46,14 @@ func Test_HTTPPoetClient_ParsesURL(t *testing.T) {
 }
 
 func Test_HTTPPoetClient_Submit(t *testing.T) {
+	var path, method string
 	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		require.Equal(t, http.MethodPost, r.Method)
+		path, method = r.URL.Path, r.Method
 		w.WriteHeader(http.StatusOK)
 
-		resp, err := protojson.Marshal(&rpcapi.SubmitResponse{})
-		require.NoError(t, err)
+		resp, _ := protojson.Marshal(&rpcapi.SubmitResponse{})
 
 		w.Write(resp)
-		require.Equal(t, "/v1/submit", r.URL.Path)
 	}))
```

Comment:
Actually, the linter complains that a code, which might cause `t.Fail()` **must not** be called on a goroutine (other than the main test goroutine) because it can cause weird behaviour and difficult debug. Details: https://github.com/Antonboom/testifylint?tab=readme-ov-file#go-require

Reply:
yes, but if the changed code fails because the handler wasn't called its also difficult to debug what went wrong 😉 

I don't say the code was fine before, I'm just suggesting a different approach for solving the linter warning 🙂

Reply:
TBH, I'm not sure. This way it's less clear what is asserted in the test and it's not even guaranteed that the test would fail if any of those assertions failed. What if the client didn't handle the error at all?

Reply:
Then the client isn't implemented correctly. The `httptest.Server` is basically a mock and we assert that our implementation behaves correctly if the mock behaves a certain way. In the example the mock returns the expected response when the client makes a POST to `/v1/submit` and otherwise 400 if the client did something unexpected and 500 if something went wrong in the mock.

Instead of asserting that the client did specific requests to the mock, we are asserting that the client behaves a correctly if the mock does.

Reply:
>Instead of asserting that the client did specific requests to the mock, we are asserting that the client behaves correctly if the mock does.

The point is to check if the client makes specific requests.

Reply:
Depends... Arguably it doesn't matter - if it made a different request the test would still fail. Because we expect the client to return a response that is based on what the server responded to its request.

To me there is little difference between:

```go
var method string
ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		method = r.Method
		w.WriteHeader(http.StatusOK)
})
// make request
require.Equal(t, http.MethodPost, r.Method)
```

and

```go
ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		require.Equal(t, http.MethodPost, r.Method)
		w.WriteHeader(http.StatusOK)
})
// make request
```

To quote `testifylint` for why it prints a warning in the second case:

```
In addition, the checker warns about require in HTTP handlers (functions and methods whose signature matches [http.HandlerFunc](https://pkg.go.dev/net/http#HandlerFunc)), because handlers run in a separate [service goroutine](https://cs.opensource.google/go/go/+/refs/tags/go1.22.3:src/net/http/server.go;l=2782;drc=1d45a7ef560a76318ed59dfdb178cecd58caf948) that services the HTTP connection. Terminating these goroutines can lead to undefined behaviour and difficulty debugging tests. You can turn off the check using the --go-require.ignore-http-handlers flag.

P.S. Look at https://github.com/stretchr/testify/issues/772, related to assertions in the goroutines.
```

And the suggested change doesn't address the underlying problem - it just shifts a possible race on `t` to a possible race on `method`.

Reply:
> but there is still the issue that a variable without mutex protection is modified on two different go routines.

It's modified on 1 goroutine and read on a second. The accesses are synchronized by the fact that the request completes by the time the variable is read so there is no race.

Reply:
@fasmat updated, please check again